src/rust/engine/process_execution/src/docker.rs

```diff
@@ -295,222 +352,331 @@ impl super::CommandRunner for CommandRunner {
+      cmd: Some(req.argv),
+      working_dir: Some(working_dir),
+      attach_stdout: Some(true),
+      attach_stderr: Some(true),
+      ..bollard::exec::CreateExecOptions::default()
+    };
+
+    log::trace!("creating execution with config: {:?}", &config);
+
+    let exec = docker
+      .create_exec::<String>(&container_id, config)
+      .await
+      .map_err(|err| format!("Failed to create Docker execution in container: {:?}", err))?;
+
+    log::trace!("created execution {}", &exec.id);
+
+    let exec_result = docker
+      .start_exec(&exec.id, None)
+      .await
+      .map_err(|err| format!("Failed to start Docker execution `{}`: {:?}", &exec.id, err))?;
+    let mut output_stream = if let StartExecResults::Attached { output, .. } = exec_result {
+      output.boxed()
+    } else {
+      panic!(
+        "Unexpected value returned from start_exec: {:?}",
+        exec_result
+      );
+    };
+
+    log::trace!("started execution {}", &exec.id);
+
+    let exec_id = exec.id.to_owned();
+    let docker = docker.clone();
+
+    let stream = async_stream::try_stream! {
+      // Read output from the execution.
+      while let Some(output_msg) = output_stream.next().await {
+        match output_msg {
+            Ok(LogOutput::StdOut { message }) => {
+                log::trace!("execution {} wrote {} bytes to stdout", &exec_id, message.len());
+                yield ChildOutput::Stdout(message);
+            }
+            Ok(LogOutput::StdErr { message }) => {
+                log::trace!("execution {} wrote {} bytes to stderr", &exec_id, message.len());
+                yield ChildOutput::Stderr(message);
+            }
+            Ok(_) => (),
+            Err(err) => {
+                log::trace!("error while capturing output of execution {}: {:?}", &exec_id, err);
+            }
+        }
+      }
+
+      let exec_metadata = docker
+        .inspect_exec(&exec_id)
+        .await
+        .map_err(|err| format!("Failed to inspect Docker execution `{}`: {:?}", &exec_id, err))?;
+
+      let status_code = exec_metadata
+        .exit_code
+        .ok_or_else(|| format!("Inspected execution `{}` for exit status but status was missing.", &exec_id))?;
+
+      log::trace!("execution {} exited with status code {}", &exec_id, status_code);
+
+      yield ChildOutput::Exit(ExitCode(status_code as i32));
+    };
+
+    Ok(stream.boxed())
+  }
+}
+
+/// Caches running containers so that build actions can be invoked by running "executions"
+/// within those cached containers.
+struct ContainerCache {
+  docker: DockerOnceCell,
+  work_dir_base: String,
+  named_caches_base_dir: String,
+  immutable_inputs_base_dir: String,
+  image_pull_cache: Arc<ImagePullCache>,
+  /// Cache that maps image name to container ID. async_oncecell::OnceCell is used so that
+  /// multiple tasks trying to access an initializing container do not try to start multiple
+  /// containers.
+  containers: Mutex<BTreeMap<String, Arc<OnceCell<String>>>>,
```

Comment:
It would probably be worth having `type` aliases for these two uses of `String`, since they are used elsewhere as well.